File: frontend/src/types/course.ts
// Types derived from backend Pydantic schemas:
//   backend/app/schemas/course.py
//   backend/app/models/course.py

// ── API Response Types (match backend exactly) ──────────────────────────

/** Backend: CourseResponse – GET /courses, GET /courses/:id */
export interface CourseResponse {
  id: string;
  code: string | null;
  subject: string;
  name: string;
  description: string | null;
  target_exam: string | null;
  estimated_hours: number | null;
  is_published: boolean;
}

/** Backend: UnitResponse – GET /courses/:id/units */
export interface UnitResponse {
  id: string;
  course_id: string;
  code: string | null;
  name: string;
  description: string | null;
  sort_order: number;
  required_mastery: number;
}

/** Backend: LessonResponse – GET /courses/:id/lessons, GET /lessons/:id */
export interface LessonResponse {
  id: string;
  unit_id: string;
  knowledge_point_id: string | null;
  code: string | null;
  title: string;
  lesson_type: string | null;
  estimated_minutes: number | null;
  sort_order: number;
  is_published: boolean;
  content: Record<string, unknown> | null;
}

// ── Backward-compatible aliases ─────────────────────────────────────────

export type Course = CourseResponse;
export type Unit = UnitResponse;
export type Lesson = LessonResponse;

// ── Composite / Frontend-only types ─────────────────────────────────────

export interface UnitWithLessons extends Unit {
  lessons: Lesson[];
}

export interface PracticeProblem {
  question: string;
  options: string[];
  answer: string;
  problem_id?: string;
}

export interface LessonSection {
  type: 'introduction' | 'concept' | 'example' | 'practice' | 'summary' | 'animation'
    | 'text' | 'formula' | 'expandable' | 'interactive_table' | 'voice_input' | 'illustration'
    | 'audio' | 'image' | 'geogebra' | 'divider' | 'code';
  title?: string;
  content?: string;
  variant?: string;
  problem?: string;
  solution?: string;
  problems?: PracticeProblem[];
  animationUrl?: string;
  animationType?: 'manim' | 'lottie' | 'css' | 'canvas';
  thumbnailUrl?: string;
  durationSec?: number;
  note?: string;
  /** For interactive_table blocks */
  tableHeaders?: string[];
  tableRows?: string[][];
  tableAnswerRows?: string[][];
  /** For expandable blocks */
  expandableItems?: { type: string; content: string; variant?: string }[];
  /** For voice_input blocks */
  voicePrompt?: string;
  /** Phase label from step */
  phase?: string;
  /** For audio blocks */
  audioUrl?: string;
  audioDuration?: number;
  audioTranscript?: string;
  audioLabel?: string;
  audioAutoplay?: boolean;
  /** For image blocks */
  imageUrl?: string;
  imageAlt?: string;
  imageCaption?: string;
  /** For geogebra blocks */
  geogebraMaterialId?: string;
  geogebraInstructions?: string;
  geogebraWidth?: number;
  geogebraHeight?: number;
  /** For divider blocks */
  dividerVariant?: 'line' | 'spacing' | 'dots' | 'label';
  dividerLabel?: string;
  /** For code blocks */
  codeContent?: string;
  codeLanguage?: string;
}

/** Enriched lesson detail returned by GET /lessons/:id */
export interface LessonDetailResponse {
  id: string;
  title: string;
  lesson_type: string | null;
  estimated_minutes: number | null;
  content: { sections: LessonSection[] };
  unit_id: string;
  unit_title: string;
  course_id: string;
  course_name: string;
  prev_lesson_id: string | null;
  next_lesson_id: string | null;
  is_enrolled: boolean;
  status: 'locked' | 'in_progress' | 'completed';
}

export interface LessonProgressResponse {
  message: string;
  progress: number;
  xp_earned: number;
}

File: backend/app/models/course.py
from uuid import UUID, uuid4

from sqlalchemy import Boolean, ForeignKey, SmallInteger, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Course(Base):
    __tablename__ = "courses"

    id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid4
    )
    code: Mapped[str | None] = mapped_column(String(50), unique=True, nullable=True)
    subject: Mapped[str] = mapped_column(String(20), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    target_exam: Mapped[str | None] = mapped_column(String(50), nullable=True)
    estimated_hours: Mapped[int | None] = mapped_column(
        SmallInteger, nullable=True
    )
    is_published: Mapped[bool] = mapped_column(Boolean, default=False)

    units: Mapped[list["Unit"]] = relationship("Unit", back_populates="course")


class Unit(Base):
    __tablename__ = "units"

    id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid4
    )
    course_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("courses.id")
    )
    code: Mapped[str | None] = mapped_column(String(50), nullable=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    sort_order: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    required_mastery: Mapped[float] = mapped_column(default=0.8)

    course: Mapped["Course"] = relationship("Course", back_populates="units")
    lessons: Mapped[list["Lesson"]] = relationship("Lesson", back_populates="unit")


class Lesson(Base):
    __tablename__ = "lessons"

    id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid4
    )
    unit_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("units.id")
    )
    knowledge_point_id: Mapped[UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("knowledge_points.id"), nullable=True
    )
    code: Mapped[str | None] = mapped_column(String(50), nullable=True)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    lesson_type: Mapped[str | None] = mapped_column(String(30), nullable=True)
    content: Mapped[dict] = mapped_column(JSONB, nullable=False)
    estimated_minutes: Mapped[int | None] = mapped_column(
        SmallInteger, nullable=True
    )
    sort_order: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    is_published: Mapped[bool] = mapped_column(Boolean, default=False)

    unit: Mapped["Unit"] = relationship("Unit", back_populates="lessons")
